clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java

```diff
@@ -639,56 +835,175 @@ public void assign(Collection<TopicPartition> partitions) {
+        } catch (InterruptException e) {
+            log.trace("Timeout during fetch", e);
+        } finally {
+            timer.update(pollTimer.currentTimeMs());
+        }
+
+        return collectFetch();
+    }
+
+    /**
+     * Perform the "{@link FetchCollector#collectFetch(FetchBuffer) fetch collection}" step by reading raw data out
+     * of the {@link #fetchBuffer}, converting it to a well-formed {@link CompletedFetch}, validating that it and
+     * the internal {@link SubscriptionState state} are correct, and then converting it all into a {@link Fetch}
+     * for returning.
+     *
+     * <p/>
+     *
+     * This method will {@link ApplicationEventHandler#wakeupNetworkThread() wake up} the {@link ConsumerNetworkThread} before
+     * retuning. This is done as an optimization so that the <em>next round of data can be pre-fetched</em>.
+     */
+    private Fetch<K, V> collectFetch() {
+        final Fetch<K, V> fetch = fetchCollector.collectFetch(fetchBuffer);
+
+        // Notify the network thread to wake up and start the next round of fetching.
+        applicationEventHandler.wakeupNetworkThread();
+
+        return fetch;
+    }
+    /**
+     * Set the fetch position to the committed position (if there is one)
+     * or reset it using the offset reset policy the user has configured.
+     *
+     * @throws org.apache.kafka.common.errors.AuthenticationException if authentication fails. See the exception for more details
+     * @throws NoOffsetForPartitionException If no offset is stored for a given partition and no offset reset policy is
+     *             defined
+     * @return true iff the operation completed without timing out
+     */
+    private boolean updateFetchPositions(final Timer timer) {
+        // Validate positions using the partition leader end offsets, to detect if any partition
+        // has been truncated due to a leader change. This will trigger an OffsetForLeaderEpoch
+        // request, retrieve the partition end offsets, and validate the current position against it.
+        applicationEventHandler.add(new ValidatePositionsApplicationEvent());
+
+        cachedSubscriptionHasAllFetchPositions = subscriptions.hasAllFetchPositions();
+        if (cachedSubscriptionHasAllFetchPositions) return true;
+
+        // Reset positions using committed offsets retrieved from the group coordinator, for any
+        // partitions which do not have a valid position and are not awaiting reset. This will
+        // trigger an OffsetFetch request and update positions with the offsets retrieved. This
+        // will only do a coordinator lookup if there are partitions which have missing
+        // positions, so a consumer with manually assigned partitions can avoid a coordinator
+        // dependence by always ensuring that assigned partitions have an initial position.
+        if (isCommittedOffsetsManagementEnabled() && !refreshCommittedOffsetsIfNeeded(timer))
+            return false;
```

Comment:
Would it be clearer to rename `refreshCommittedOffsetsIfNeeded` to `initWithCommittedOffsetsIfNeeded`?

Reply:
The `refreshCommittedOffsetsIfNeeded()` name is derived from existing code in `KafkaConsumer` that calls out to `ConsumerCoordinator.refreshCommittedOffsetsIfNeeded()`. If we change the method name in `PrototypeAsyncConsumer`, should we change `ConsumerCoordinator`'s corresponding method name too?

Reply:
Yes, in both cases, the method sets offsets for partitions in initializing state.

Reply:
Updated and a bit of cleanup.